### ISSUE TITLE
[🐛] Bug Report Title -🔥  Crashlytics unable to detect the crashes in iOS app

### ISSUE THREAD

**Issue by Yandamuri** [invertase/react-native-firebase#6614]
## Issue

I have configured `@react-native-firebase/crashlytics` in my project and able to crash the app with following,

    crashlytics().crash();

App crash is successful but it was not detectable by `Firebase Crashlytics`. Crashlytics dashboard page keep on showing that `App detected and we're waiting for a crash!`.

Please find the reproducible demo here -- https://github.com/Yandamuri/react-native-crash-test

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

       
```json
{
  "name": "sample",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
  },
  "dependencies": {
    "@react-native-firebase/app": "^15.7.0",
    "@react-native-firebase/crashlytics": "^15.7.0",
    "expo": "^46.0.0",
    "react": "18.1.0",
    "react-native": "0.70.1",
    "react-native-gesture-handler": "^2.6.1",
    "react-native-reanimated": "^2.10.0"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.5",
    "@react-native-community/eslint-config": "^2.0.0",
    "babel-jest": "^26.6.3",
    "eslint": "^7.32.0",
    "jest": "^26.6.3",
    "metro-react-native-babel-preset": "^0.72.1",
    "react-test-renderer": "18.1.0"
  },
  "jest": {
    "preset": "react-native"
  }
}

```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
$RNFirebaseAsStaticFramework = true;

require File.join(File.dirname(`node --print "require.resolve('expo/package.json')"`), "scripts/autolinking")
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '12.4'
install! 'cocoapods', :deterministic_uuids => false

target 'sample' do
  use_expo_modules!
  post_integrate do |installer|
    begin
      expo_patch_react_imports!(installer)
    rescue => e
      Pod::UI.warn e
    end
  end
  config = use_native_modules!

  # Flags change depending on the env values.
  flags = get_default_flags()

  use_react_native!(
    :path => config[:reactNativePath],
    # Hermes is now enabled by default. Disable by setting this flag to false.
    # Upcoming versions of React Native may rely on get_default_flags(), but
    # we make it explicit here to aid in the React Native upgrade process.
    :hermes_enabled => true,
    :fabric_enabled => flags[:fabric_enabled],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    :flipper_configuration => FlipperConfiguration.enabled,
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

    pod 'Firebase', :modular_headers => true
    pod 'FirebaseCoreInternal', :modular_headers => true
    pod 'GoogleUtilities/Environment', :modular_headers => true
    pod 'FirebaseCore', :modular_headers => true
    

  target 'sampleTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    use_frameworks!
    react_native_post_install(
      installer,
      # Set `mac_catalyst_enabled` to `true` in order to apply patches
      # necessary for Mac Catalyst builds
      :mac_catalyst_enabled => false
    )
    __apply_Xcode_12_5_M1_post_install_workaround(installer)
  end
end

```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import <Firebase.h>

#import <React/RCTAppSetupUtils.h>

#if RCT_NEW_ARCH_ENABLED
#import <React/CoreModulesPlugins.h>
#import <React/RCTCxxBridgeDelegate.h>
#import <React/RCTFabricSurfaceHostingProxyRootView.h>
#import <React/RCTSurfacePresenter.h>
#import <React/RCTSurfacePresenterBridgeAdapter.h>
#import <ReactCommon/RCTTurboModuleManager.h>

#import <react/config/ReactNativeConfig.h>

static NSString *const kRNConcurrentRoot = @"concurrentRoot";

@interface AppDelegate () <RCTCxxBridgeDelegate, RCTTurboModuleManagerDelegate> {
  RCTTurboModuleManager *_turboModuleManager;
  RCTSurfacePresenterBridgeAdapter *_bridgeAdapter;
  std::shared_ptr<const facebook::react::ReactNativeConfig> _reactNativeConfig;
  facebook::react::ContextContainer::Shared _contextContainer;
}
@end
#endif

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];  
  RCTAppSetupPrepareApp(application);

  RCTBridge *bridge = [self.reactDelegate createBridgeWithDelegate:self launchOptions:launchOptions];

#if RCT_NEW_ARCH_ENABLED
  _contextContainer = std::make_shared<facebook::react::ContextContainer const>();
  _reactNativeConfig = std::make_shared<facebook::react::EmptyReactNativeConfig const>();
  _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
  _bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:bridge contextContainer:_contextContainer];
  bridge.surfacePresenter = _bridgeAdapter.surfacePresenter;
#endif

  NSDictionary *initProps = [self prepareInitialProps];
  UIView *rootView = RCTAppSetupDefaultRootView(bridge, @"sample", initProps);

  if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [self.reactDelegate createRootViewController];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [super application:application didFinishLaunchingWithOptions:launchOptions];
  return YES;
}

/// This method controls whether the `concurrentRoot`feature of React18 is turned on or off.
///
/// @see: https://reactjs.org/blog/2022/03/29/react-v18.html
/// @note: This requires to be rendering on Fabric (i.e. on the New Architecture).
/// @return: `true` if the `concurrentRoot` feture is enabled. Otherwise, it returns `false`.
- (BOOL)concurrentRootEnabled
{
  // Switch this bool to turn on and off the concurrent root
  return true;
}

- (NSDictionary *)prepareInitialProps
{
  NSMutableDictionary *initProps = [NSMutableDictionary new];

#ifdef RCT_NEW_ARCH_ENABLED
  initProps[kRNConcurrentRoot] = @([self concurrentRootEnabled]);
#endif

  return initProps;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

#if RCT_NEW_ARCH_ENABLED

#pragma mark - RCTCxxBridgeDelegate

- (std::unique_ptr<facebook::react::JSExecutorFactory>)jsExecutorFactoryForBridge:(RCTBridge *)bridge
{
  _turboModuleManager = [[RCTTurboModuleManager alloc] initWithBridge:bridge
                                                             delegate:self
                                                            jsInvoker:bridge.jsCallInvoker];
  return RCTAppSetupDefaultJsExecutorFactory(bridge, _turboModuleManager);
}

#pragma mark RCTTurboModuleManagerDelegate

- (Class)getModuleClassFromName:(const char *)name
{
  return RCTCoreModulesClassProvider(name);
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                      jsInvoker:(std::shared_ptr<facebook::react::CallInvoker>)jsInvoker
{
  return nullptr;
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:(const std::string &)name
                                                     initParams:
                                                         (const facebook::react::ObjCTurboModule::InitParams &)params
{
  return nullptr;
}

- (id<RCTTurboModule>)getModuleInstanceFromClass:(Class)moduleClass
{
  return RCTAppSetupDefaultModuleFromClass(moduleClass);
}

#endif

@end

```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
 System:
    OS: macOS 12.5
    CPU: (8) x64 Apple M1
    Memory: 45.83 MB / 8.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 14.0.0 - /usr/local/bin/node
    Yarn: Not Found
    npm: 6.14.4 - /usr/local/bin/npm
    Watchman: Not Found
  Managers:
    CocoaPods: 1.11.2 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK: Not Found
  IDEs:
    Android Studio: 2021.3 AI-213.7172.25.2113.9014738
    Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.10 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.1.0 => 18.1.0 
    react-native: 0.70.1 => 0.70.1 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [x ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `15.7.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `crashlytics`
- **Are you using `TypeScript`?**
  - `N`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
I've written a whole article on this: https://invertase.io/blog/react-native-firebase-crashlytics-configuration

Reading that guide, you may watch the logs to determine exactly how it is configured at app boot, and to watch crash behavior.
I can assure you it's working, though sometimes crashes take a little while to show up and for the console to let you in to see them.

Make sure you have selected the correct app from the crashlytics app selection drop-down (ios and android are separate, and you may have more than one app configured - meaning you are actually looking for crashes from your android app or the wrong ios app sometimes)

You must not have a debugger attached when you crash, and typically you must not be using a debug build unless configured otherwise.

Also, you must restart the app after the crash for the report to send (and if you watch the logs via console.app you should see it send)

This is obviously a vital feature so I just tested it right now again to verify, and it appears to work fine using 1) our tests app here, the "native crash now button", 2) restarting the app 3) on an iPhone 14 simulator running on macOS-current

![image](https://user-images.githubusercontent.com/782704/196722545-419d4765-77c8-4372-a252-be7d09e5f197.png)

I have not run your reproduction - I don't have time to pull it and run it right now - but hopefully the ability to see how our test app is configured (examine the [firebase.json](https://github.com/invertase/react-native-firebase/blob/fe2bebee106dc178b0835f7097c2d05b30223b2f/tests/firebase.json#L10-L15)) as well as the document I linked above is enough to go on

**Comment by Yandamuri** [comment]
@mikehardy Thank you for your response.

I have gone through your article and did the configuration once again, Then I was able to see the crash reports over Firebase Dashboard page.But I tried in sample project which created to report this issue. Then I tried in my actual project where I found some weird behaviour.

I used following line to crash the app,

    undefinedVariable.notAFunction()

The above line crashes the app in my sample app. Which means app is closed abruptly. But When used above line in my actual app, I found that app just got stuck and I was not able to perform any action. And Xcode throws an exception which shown in below picture,

<img width="776" alt="Screenshot 2022-10-20 at 5 18 27 PM" src="https://user-images.githubusercontent.com/31983789/197319031-199a2f2e-8441-44c8-8938-bf37ae9acaac.png">


I tested with different crash examples. Every time app is just getting stuck and throwing exception in `Xcode`. But no crash at all. 

Can we consider this behaviour as crash? If so Why Firebase Analytics is unable to catch it?

If it is not a crash, Then how to crash an app(I mean abruptly closing the app)?

I am sure library is working well and this is not a package issue. But Can you please help me here that why `undefinedVariable.notAFunction()` didn't crash the app, instead throwing an exception in `Xcode`?

**Comment by mikehardy** [comment]
Xcode is a debugger. If you run it under Xcode it is attached to the process and stops it on crash so you can debug it.
The logs clearly indicate that if Xcode is attached as a debugger crashlytics won't work.
The article mentions the same.

**Comment by mikehardy** [comment]
I quote:

> f you start the iOS app from Xcode, there will be a debugger attached to the app, meaning you will not see the crash reports in the Firebase web console.

So it sounds like this is all working as expected - I'm going to close this as there are no actions to take for the module

**Comment by Yandamuri** [comment]
@mikehardy 

May I know the command to run the `iOS` app in release mode from Terminal?

For android it is `react-native run-android --variant release`. Is there any equivalent for iOS app to run app ijn release mode?

**Comment by mikehardy** [comment]
`npx react-native run-ios --help` will show you, I think it's with `--configuration "Release"`

**Comment by Yandamuri** [comment]
@mikehardy 
Finally I am able to see crashes over `Firebase Crashlytics dashboard` for both iOS and Android. But I am confused two things,

1. Before using `Crashlytics` I was using Firebase for notification purpose. So I already had `GoogleService-Info.plist`. So I have proceeded with rest of the steps for `Crashlytics`.  Then there was no crash was recorded over `Firebase Crashlytics dashboard`. Then Just for the testing purpose I have downloaded `GoogleService-Info.plist` and kept it inside `iOS` folder. Then surprisingly crashes were recorded over `Firebase Crashlytics dashboard`. Then I checked for git changes in newly dowloaded `GoogleService-Info.plist`. I found following change,

```
<key>ANDROID_CLIENT_ID</key>
<string><Some String here></string>
```

   will `ANDROID_CLIENT_ID` be added to `GoogleService-Info.plist` if we configure `@react-native-firebase/crashlytics`?

I thought this `ANDROID_CLIENT_ID` is required for `Crashlytics`. But still I am doubt about it, so I reverted these changes, which means my `GoogleService-Info.plist` is like earlier. Then again I tested crashes. still it is working and crashes were recorded.

Isn't old `GoogleService-Info.plist` sufficient for to make `@react-native-firebase/crashlytics` work?

2. 

> Reports are sent at app start, so close and restart the app to send the report.

This line is from your article. This is true in case of `iOS`. But in `Android`, I am able to see crashes even without restarting the app.

May I know the reason behind it?
